src/platform/intel/cavs/platform.c

```diff
@@ -543,7 +543,47 @@ void platform_wait_for_interrupt(int level)
 }
 #endif
 
+#define ADSP_IMR_MAGIC_VALUE		0x02468ACE
+#define IMR_L1_CACHE_ADDRESS		0xB0000000
+
+struct imr_header {
+	uint32_t adsp_imr_magic;
+	uint32_t structure_version;
+	uint32_t structure_size;
+	uint32_t imr_state;
+	uint32_t imr_size;
+	void *imr_restore_vector;
+};
+
+struct imr_state {
+	struct imr_header header;
+	uint8_t reserved[0x1000 - sizeof(struct imr_header)];
+};
+
+struct irm_layout {
+	uint8_t     css_reserved[0x1000];
+	struct imr_state    imr_state;
+};
+
+static struct irm_layout *imr_layout;
+
+static void imr_layout_init(void)
+{
+	imr_layout = (struct irm_layout *)(IMR_L1_CACHE_ADDRESS);
+	xthal_dcache_region_invalidate(imr_layout, sizeof(struct irm_layout));
+}
+
+static void WriteMagicAndVector(void *vector)
+{
+	imr_layout->imr_state.header.adsp_imr_magic = ADSP_IMR_MAGIC_VALUE;
+	imr_layout->imr_state.header.imr_restore_vector = vector;
+	xthal_dcache_region_writeback(imr_layout, sizeof(struct irm_layout));
+}
+
 int platform_suspend(struct sof *sof)
 {
+	imr_layout_init();
+	WriteMagicAndVector((void *)IMR_BOOT_LDR_TEXT_ENTRY_BASE);
```

Comment:
Where does this come from - and coding style.

Reply:
> Where does this come from - and coding style.

oops, let me fix it.